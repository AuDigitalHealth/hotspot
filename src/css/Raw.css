.raw {
  font-size: 13px;
}

.raw pre {
  background-color: transparent;
}
