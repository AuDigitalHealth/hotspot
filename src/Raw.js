import React, { Component } from 'react'
import Highlight from 'react-highlight'

import 'highlight.js/styles/docco.css'

// Renders the raw content of an XML or JSON FHIR resource, pretty-printed and
// with syntax highlighting.
class Raw extends Component {
  constructor(props) {
    super(props)
    this.state = {}
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.content && nextProps.format) {
      try {
        this.setState(() => ({
          prettyContent: Raw.prettifyContent(nextProps.content, nextProps.format),
        }))
      } catch (error) {
        this.props.onError(error)
      }
    }
  }

  render() {
    const { format } = this.props
    const { prettyContent } = this.state

    return <div className='raw'>
      <Highlight className={format}>{prettyContent}</Highlight>
    </div>
  }

  static prettifyContent(content, format) {
    if (format === 'json') {
      return Raw.prettifyJSONContent(content)
    } else if (format === 'xml') {
      return Raw.prettifyXMLContent(content)
    } else {
      throw new Error('Unsupported content type.')
    }
  }

  static prettifyJSONContent(content) {
    return JSON.stringify(JSON.parse(content), null, 2)
  }

  static prettifyXMLContent(content) {
    const xslt = `<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
 <xsl:output omit-xml-declaration="yes" indent="yes"/>
    <xsl:template match="node()|@*">
      <xsl:copy>
        <xsl:apply-templates select="node()|@*"/>
      </xsl:copy>
    </xsl:template>
</xsl:stylesheet>`

    const parser = new DOMParser()
    const xsltDoc = parser.parseFromString(xslt, 'application/xml')
    const xmlDoc = parser.parseFromString(content, 'application/xml')
    const processor = new XSLTProcessor()
    processor.importStylesheet(xsltDoc)
    const transformed = processor.transformToDocument(xmlDoc)

    return new XMLSerializer().serializeToString(transformed)
  }
}

export default Raw
