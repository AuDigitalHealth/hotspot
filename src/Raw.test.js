import React from 'react'
import { shallow } from 'enzyme'
import Highlight from 'react-highlight'

import Raw from './Raw.js'

describe('Raw', () => {
  it('should render JSON correctly', () => {
    const props = {
      content: '{ "cat": [ "dog", "catdog" ] }',
      format: 'json',
      onError: jest.fn(),
    }
    const wrapper = shallow(<Raw {...props} />)
    expect(wrapper.find(Highlight).children()).toMatchSnapshot()
    expect(props.onError).not.toHaveBeenCalled()
  })

  // We can't currently test this, as XSLTProcessor is not supported within
  // JSDOM.
  // it('should render XML correctly', () => {
  //   const props = {
  //     content: '<cat><dog><catdog /></dog></cat>',
  //     format: 'xml',
  //     onError: jest.fn(),
  //   }
  //   const wrapper = mount(<Raw {...props} />)
  //   expect(wrapper.find(Highlight).children()).toMatchSnapshot()
  //   expect(props.onError).not.toHaveBeenCalled()
  // })
})
