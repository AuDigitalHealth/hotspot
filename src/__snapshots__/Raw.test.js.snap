// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raw should render JSON correctly 1`] = `
<div
  className="raw"
>
  <pre
    className="json"
  >
    {
      "cat": [
        "dog",
        "catdog"
      ]
    }
  </pre>
</div>
`;
